### PR TITLE
Don't select more inputs than allowed

### PR DESCRIPTION
**Pull request by Thoralf-M**
# Description of change

Don't select more inputs than allowed

## Links to any relevant issues

https://github.com/iotaledger/iota-sdk/issues/68

## Type of change

- Bug fix (a non-breaking change which fixes an issue)

## How the change has been tested

Added tests

## Change checklist

- [ ] I have followed the contribution guidelines for this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that new and existing unit tests pass locally with my changes

### CODE REVIEW

File: client/src/api/block_builder/input_selection/core/mod.rs [hunk 1]
```diff
@@ -390,6 +397,12 @@ impl InputSelection {
 
         self.outputs.extend(storage_deposit_returns);
 
+        if !OUTPUT_COUNT_RANGE.contains(&(self.outputs.len() as u16)) {
+            return Err(Error::Block(iota_types::block::Error::InvalidOutputCount(
+                TryIntoBoundedU16Error::Truncated(self.outputs.len()),
+            )));
```
Comment (thibault-martinez) [inline client/src/api/block_builder/input_selection/core/mod.rs]: I think I would have a dedicated variant in the ISA error tbh, it's not exactly the same error, trying to select too much vs. trying to build a block with too much. And having this packable stuff here is a bit convoluted.

File: client/CHANGELOG.md [hunk 2]
```diff
@@ -25,6 +25,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 - `SecretManager::{try_from_mnemonic, try_from_hex_seed}`;
 - Derive `Serialize` for `GetAddressesBuilderOptions`, `ClientBlockBuilderOutputAddress`, `ClientBlockBuilderOptions` and `Message`;
+- Derive `Eq, PartialEq` for `input_select::Error`;
```
Comment (thibault-martinez) [inline client/CHANGELOG.md]: Rebase and move to a new section as rc7 has been released

File: client/CHANGELOG.md [hunk 3]
```diff
@@ -52,6 +53,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - ISA does not select aliases it can't unlock for amount;
 - Regex of several MQTT topics;
 - ISA makes sure the address is owned on alias Ed25519 requirements;
+- ISA will not select more than max inputs;
```
Comment (thibault-martinez) [inline client/CHANGELOG.md]: Same

File: client/CHANGELOG.md [hunk 4]
```diff
@@ -27,6 +27,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - MQTT `Error`;
 - `Message::HashTransactionEssence`;
 - `Response::TransactionEssenceHash`;
+- Derive `Eq, PartialEq` for `input_select::Error`;
+- `Error::{InvalidOutputCount, InvalidOutputCount}` variants;
```
Comment (thibault-martinez) [inline client/CHANGELOG.md]:
```suggestion
- ISA `Error::{InvalidOutputCount, InvalidOutputCount}` variants;
```

File: client/src/api/block_builder/input_selection/core/mod.rs [hunk 5]
```diff
@@ -390,6 +395,10 @@ impl InputSelection {
 
         self.outputs.extend(storage_deposit_returns);
 
+        if !OUTPUT_COUNT_RANGE.contains(&(self.outputs.len() as u16)) {
```
Comment (thibault-martinez) [inline client/src/api/block_builder/input_selection/core/mod.rs:403]: Can you add a test for too many outputs pls?

Comment (Thoralf-M) [reply client/src/api/block_builder/input_selection/core/mod.rs:403]: Done

**Pull request closed without merging**